ddspipe_core/src/cpp/communication/dds/DdsBridge.cpp

```diff
@@ -175,6 +97,178 @@ void DdsBridge::disable() noexcept
+            {
+                // The participant is not a repeater. Do not add its writer.
+                writers_ids.erase(id);
+            }
+
+            writers_to_create.insert(writers_ids.begin(), writers_ids.end());
+        }
+    }
+
+    // Create the writers.
+    std::map<ParticipantId, std::shared_ptr<IWriter>> writers;
+
+    for (const auto& id : writers_to_create)
+    {
+        std::shared_ptr<IParticipant> participant = participants_->get_participant(id);
+        writers[id] = participant->create_writer(*topic_);
+    }
+
+    // Add the writers to the tracks they have routes for.
+    add_writers_to_tracks_(writers);
+}
+
+void DdsBridge::add_to_tracks(
+        const ParticipantId& discoverer_participant_id)
+{
+    // A new subscriber has been discovered.
+    // Check if the writer for this participant already exists.
+    for (const auto& tracks_it : tracks_)
+    {
+        const auto& track = tracks_it.second;
+
+        if (track->has_writer(discoverer_participant_id))
+        {
+            // The writer already exists. There is nothing to do. Exit.
+            return;
+        }
+    }
+
+    std::shared_ptr<IParticipant> participant = participants_->get_participant(discoverer_participant_id);
+
+    // Create the writer.
+    std::map<ParticipantId, std::shared_ptr<IWriter>> writer;
+    writer[discoverer_participant_id] = participant->create_writer(*topic_);
+
+    // Add the writer to the tracks it has routes for.
+    add_writers_to_tracks_(writer);
+}
+
+void DdsBridge::remove_from_tracks(
+        const ParticipantId& discoverer_participant_id) noexcept
+{
+    for (auto it = tracks_.cbegin(), next_it = it; it != tracks_.cend(); it = next_it)
+    {
+        ++next_it;
+
+        const auto& track = it->second;
+
+        // If the writer is in the track, remove it.
+        track->remove_writer(discoverer_participant_id);
+
+        if (track->count_writers() <= 0)
+        {
+            // The track doesn't have any writers. Remove it.
+            tracks_.erase(it);
+        }
+    }
+}
+
+void DdsBridge::add_writers_to_tracks_(
+        std::map<ParticipantId, std::shared_ptr<IWriter>>& writers)
+{
+    // Add writers to the tracks of the readers in their route.
+    // If the readers in their route don't exist, create them with their tracks.
+    for (const ParticipantId& id : participants_->get_participants_ids())
+    {
+
+        // Select the necessary writers
+        std::map<ParticipantId, std::shared_ptr<IWriter>> writers_of_reader;
+
+        const auto& routes_it = routes_.find(id);
+
+        if (routes_it != routes_.end())
+        {
+            // The reader has a route. Add only the writers in the route.
+            const auto& writers_in_route_of_reader = routes_it->second;
+
+            for (const auto& writer_id : writers_in_route_of_reader)
+            {
+                writers_of_reader[writer_id] = writers[writer_id];
```

Comment:
I think this is wrong, you may end up dealing with nullpointers in some situations.